waiter/src/waiter/statsd.clj

```diff
@@ -375,53 +375,51 @@
-    (let [counts-by-metric-group'
-          (cond-> counts-by-metric-group
-                  (= message-type :update-service-instances)
-                  (process-update-service-instances-message message-data service-id->service-description-fn))]
-      (if (some? remaining)
-        (recur remaining counts-by-metric-group')
-        counts-by-metric-group'))))
-
-(defn process-scheduler-messages
+      (log/error e "error processing service metrics for" service-id)
+      metric-group->counts)))
+
+(defn router-state->metric-group->counts
+  "Converts messages from the router state update to a map of metric-group -> [healthy unhealthy failed],
+   where each element in the array is the count of instances in that metric group with that status."
+  [service-id->service-description-fn
+   {:keys [all-available-service-ids service-id->failed-instances service-id->healthy-instances service-id->unhealthy-instances]}]
+  (loop [[service-id & remaining-service-ids] (seq all-available-service-ids)
+         metric-group->counts {}]
+    (if-not service-id
+      metric-group->counts
+      (let [failed-instances (get service-id->failed-instances service-id)
+            healthy-instances (get service-id->healthy-instances service-id)
+            unhealthy-instances (get service-id->unhealthy-instances service-id)]
+        (recur remaining-service-ids
+               (process-service-instance-state
+                 service-id->service-description-fn service-id healthy-instances unhealthy-instances failed-instances
+                 metric-group->counts))))))
+
+(defn process-router-state
   "Publishes all gauges produced from the provided scheduler messages"
-  [messages service-id->service-description-fn]
+  [router-state service-id->service-description-fn]
   (try
-    (->
-      messages
-      (scheduler-messages->instance-counts-by-metric-group service-id->service-description-fn)
-      (publish-instance-counts-by-metric-group))
+    (->> router-state
+         (router-state->metric-group->counts service-id->service-description-fn)
+         publish-metric-group->counts)
     (catch Throwable e
-      (log/error e "Error processing scheduler messages"))))
-
-(defn start-scheduler-metrics-publisher
-  "Go loop to continuously publish stats based on scheduler messages"
-  [scheduler-state-chan exit-chan service-id->service-description-fn]
-  (log/info "Starting scheduler-metrics-publisher")
-  (async/go-loop []
-    (let [continue
-          (async/alt!
-            exit-chan
-            ([_] false)
-
-            scheduler-state-chan
-            ([messages]
-              (process-scheduler-messages messages service-id->service-description-fn)
-              true))]
-      (if continue
-        (recur)
-        (log/info "Stopping scheduler-metrics-publisher")))))
+      (log/error e "error processing router state"))))
+
+(defn start-service-instance-metrics-publisher
+  "Go loop to continuously publish stats based on router state updates."
```

Comment:
This is no longer explicitly a `go` loop

Reply:
Updated comment